scripts/documentation/tsdoc.ts

```diff
@@ -0,0 +1,333 @@
+  events: ModuleEvent[];
+}
+
+interface ModuleEvent {
+  documentation: string;
+}
+
+interface ModuleMethod {
+  methodSignature: string;
+  documentation: string;
+}
+
+interface ModuleProperty {
+  name: string;
+  type: string;
+  documentation: string;
+}
+
+interface DocumentationContent {
+  tag?: string;
+  value?: string;
+}
+
+const FOUNDATION = 'foundation';
+const ADAPTER = 'adapter';
+const README_FILE = 'README.md';
+
+class TypeScriptDocumentationGenerator {
+  markdownBuffer: {[s: string]: ModuleMarkdown[]};
+  docData?: {}; // Documentalist representation of methods/properties/events
+  templateFunction: Handlebars.TemplateDelegate<{}>;
+
+  constructor() {
+    this.docData = {};
+    this.markdownBuffer = {};
+    fs.readFile('./scripts/documentation/apiMarkdownTableTemplate.hbs', 'utf8', (error, template) => {
+      if (error) {
+        console.error(error); // tslint:disable-line
+      }
+      this.templateFunction = Handlebars.compile(template, {noEscape: true});
+    });
+  }
+
+  /**
+   * Generates JSON from source files TypeScript documentation.
+   * This contains all the esmodule classes (ie. foundations, adapters, components) in JSON format.
+   * @returns Promise<{}>
+   */
+  generateJSONFromFiles() {
+    return new Promise((resolve) => {
+      new Documentalist()
+        .use(/\.ts$/, new TypescriptPlugin({
+          excludePaths: ['node_modules'],
+          includeDeclarations: true,
+        }))
+        .documentGlobs('packages/**/*') // ← async operation, returns a Promise
+        .then((docs) => resolve(docs))
+        .catch((error) => console.error(error)); // tslint:disable-line
+    });
+  }
+
+  /**
+   * The main function of this class. Iterates through all modules found in docData (json).
+   * This should already be generated from this.generateJSONFromFiles().
+   * @param docData json containing documentation from documentalist
+   */
+  generateDocs(docData) {
+    this.docData = docData.typescript;
+    Object.keys(this.docData).forEach((module) => {
+      console.log(`-- generating docs for ${module}`); // tslint:disable-line
+      this.generateDocsForModule(module);
+    });
+
+    this.generateMarkdownFileFromBuffer();
+  }
+
+  /**
+   * Creates documentation for a specified `esmodule`.
+   * @param esmodule module name (ie. MDCSelectIconFoundation)
+   */
+  generateDocsForModule(esmodule: string) {
+    const {kind, fileName} = this.docData[esmodule];
+    const readmeDirectoryPath = this.getFilePathName(fileName);
+
+    if (kind === 'type alias') {
+      // ignore types and interfaces
+      return;
+    }
+    const markdownObject: ModuleMarkdown = {
+      events: this.getDocumentationForModule(esmodule).events,
+      methods: this.getDocumentationForMethods(esmodule),
+      moduleName: esmodule,
+      properties: this.getDocumentationProperties(esmodule),
+      readmeDirectoryPath,
+    };
+    this.addToMarkdownBuffer(readmeDirectoryPath, markdownObject);
```

Comment:
It is hard to reason about when buffer is used (side-effect), can we make this method return generated object which can be written into the README instead?

Reply:
updated to remove the markdownBuffer object from the class